### ISSUE TITLE
RBAC Principal Discovery Service Proposal 

### ISSUE THREAD

**Issue by Shikugawa** [envoyproxy/envoy#36845]
*Title*: RBAC Filter Principal Discovery Service

*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

We deploy IP-based access control using RBAC filter, and write IP CIDR by hand directly.
But, we need any other solutions to switch the set of CIDRs dynamically. Instead of writing IP CIDR in the config, we intend to write any identifier in the config, and resolve the set of CIDRs from them dynamically.

I guess currently Envoy doesn't provide such a feature. Therefore we need to invent something new to achieve this requirement.

I came up with the following solutions, I'd like to discuss for the final decision:

1. Add new xDS service RBAC Principal Discovery Service (RPDS) to retrieve the set of IP CIDRs from DiscoveryResponse. This solution is not only focusing on our requirements, but also any other dynamic principal configuration as well.
2. Configuration based on DNS. It simply loads the  set of address periodically from DNS. and configure them. But, it couldn't support CIDR-based configuration, thus it doesn't meet with our requirements.

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by nezdolik** [comment]
cc @yangminzhu @yanavlasov

**Comment by kyessenov** [comment]
Did you consider ECDS for RBAC?

**Comment by Shikugawa** [comment]
@kyessenov Thank you. I have considered to use ECDS. But I'm not sure whether it meets with our use cases.
We deploy envoy for multi-tenants. It means that there are a lot of listeners (>500), and they owns different RBAC configuration.
Therefore we need something identification like route_config_name in RDS. But I couldn't catch ECDS owns such a mechanism to apply the each different config.

Actually, what we want to configure dynamically is just allowed_source_ip and denied_source_ip in RBAC principal as well.
So having dynamic configuration mechanism like CIDR Group Discovery Service meets 100% of our requirements.

In summary, our requirements are like the following:

- We need identification mechanism for ECDS.
- We need CIDR group dynamic configuration mechanism.

**Comment by kyessenov** [comment]
ECDS resources are named, so you could attach individual RBAC config to each listener. But you are right, this would potentially increase the size of configs since there is little reuse between RBAC configuration. Another approach would be to create a new (listener?) filter, configured via ECDS, that writes a filter state based on CIDR membership. Then in RBAC you use the filter state rather than the IP directly. Creating a new xDS is complex, so re-using ECDS with a new filter might be the easiest way to accomplish what you want.

**Comment by Shikugawa** [comment]
@kyessenov Thank you. It sounds reasonable to achieve our requirements. 
But I'm interested in whether my requirement is common for Envoy users or not. I guess many users might apply ACL policies by identifier, not the set of IP CIDRs directly, and the contents of the set is updated without changing of identifier.
In this case, we want to use something dynamic resolution mechanism of the IP CIDRs from identifier.
So, I thought we should have any unified way to achive this requirement. What do you think about it?

IMO if this requirement is not so common, I think we create new listener filter to update filter_state and do the action based on them in RBAC filter. But if it's common, I guess we should prepare something straightforward approach to achieve this (IP Group Discovery Service is just one of the many approaches)
